crates/network-scanner/src/task_utils.rs

```diff
@@ -0,0 +1,153 @@
+#[derive(Debug)]
+pub(crate) struct TaskExecutionRunner {
+    pub(crate) context: TaskExecutionContext,
+    pub(crate) handles_receiver: HandlesReceiver,
+    pub(crate) handles_sender: HandlesSender,
+}
+
+impl TaskExecutionContext {
+    pub(crate) fn new(network_scanner: NetworkScanner, handles_sender: HandlesSender) -> anyhow::Result<Self> {
+        let (ip_sender, ip_receiver) = tokio::sync::mpsc::channel(1024);
+        let ip_receiver = Arc::new(Mutex::new(ip_receiver));
+
+        let (port_sender, port_receiver) = tokio::sync::mpsc::channel(1024);
+        let port_receiver = Arc::new(Mutex::new(port_receiver));
+
+        let subnets = get_subnets()?;
+        let NetworkScanner {
+            ports,
+            ping_timeout,
+            ping_interval,
+            broadcast_timeout,
+            port_scan_timeout,
+            netbios_timeout,
+            runtime,
+            ..
+        } = network_scanner;
+
+        let res = Self {
+            handles_sender,
+            ip_sender,
+            ip_receiver,
+            port_sender,
+            port_receiver,
+            ip_cache: Arc::new(DashMap::new()),
+            ports,
+            runtime,
+            ping_interval,
+            ping_timeout,
+            broadcast_timeout,
+            port_scan_timeout,
+            netbios_timeout,
+            subnets,
+        };
+
+        Ok(res)
+    }
+}
+
+impl TaskExecutionRunner {
+    // Move the generic parameters to the method level.
+    pub(crate) fn run<T, F>(&mut self, task: T)
+    where
+        T: FnOnce(TaskExecutionContext) -> F + Send + 'static,
+        F: Future<Output = anyhow::Result<()>> + Send + 'static,
+    {
+        let context = self.context.clone();
+        let handle = tokio::task::spawn(async move { task(context).await });
+        let _ = self.handles_sender.send(handle);
+    }
+
+    pub(crate) fn new(scanner: NetworkScanner) -> anyhow::Result<Self> {
+        let (handles_sender, handles_receiver) = crossbeam::channel::unbounded();
+        Ok(Self {
+            context: TaskExecutionContext::new(scanner, handles_sender.clone())?,
+            handles_receiver,
+            handles_sender,
+        })
+    }
+}
+
+pub fn clone2<A, B>(a: &A, b: &B) -> (A, B)
+where
+    A: Clone,
+    B: Clone,
+{
+    (a.clone(), b.clone())
+}
+
+pub fn clone3<A, B, C>(a: &A, b: &B, c: &C) -> (A, B, C)
+where
+    A: Clone,
+    B: Clone,
+    C: Clone,
+{
+    (a.clone(), b.clone(), c.clone())
+}
+
+pub fn clone4<A, B, C, D>(a: &A, b: &B, c: &C, d: &D) -> (A, B, C, D)
+where
+    A: Clone,
+    B: Clone,
+    C: Clone,
+    D: Clone,
+{
+    (a.clone(), b.clone(), c.clone(), d.clone())
+}
+
+pub fn clone5<A, B, C, D, E>(a: &A, b: &B, c: &C, d: &D, e: &E) -> (A, B, C, D, E)
+where
+    A: Clone,
+    B: Clone,
+    C: Clone,
+    D: Clone,
+    E: Clone,
+{
+    (a.clone(), b.clone(), c.clone(), d.clone(), e.clone())
+}
```

Comment:
nitpick: I don’t think this is very idiomatic. We typically just use `.clone()` directly explicitly. I don’t really want to adopt this style if we’re not going to be consistent.

Reply:
I understand I hesitated when I wrote this code, but it really helps with readability, there are just too many to clone when we need to use these variables in different tasks. using an extra 5 lines is really bad for readability, considering there are already a lot of indentations and logic flooding in the tasks.

I can add #[inline] tag so it would compile to the same assembly code

Reply:
updated